packages/runtime/container-runtime/src/gc/garbageCollection.ts

```diff
@@ -258,36 +255,47 @@ export class GarbageCollector implements IGarbageCollector {
 		 * Set up the initializer which initializes the GC state from the data in base snapshot. This is done when
 		 * connected in write mode or when GC runs the first time. It sets up all unreferenced nodes from the base
 		 * GC state and updates their inactive or sweep-ready state.
+		 *
+		 * Must only be called if there is a current reference timestamp.
 		 */
 		this.initializeGCStateFromBaseSnapshotP = new LazyPromise<void>(async () => {
-			/**
-			 * If there is no current reference timestamp, skip initialization. We need the current timestamp to track
-			 * how long objects have been unreferenced and if they can be deleted.
-			 *
-			 * Note that the only scenario where there is no reference timestamp is when no ops have ever been processed
-			 * for this container and it is in read mode. In this scenario, there is no point in running GC anyway
-			 * because references in the container do not change without any ops, i.e., there is nothing to collect.
-			 */
 			const currentReferenceTimestampMs = this.runtime.getCurrentReferenceTimestampMs();
-			if (currentReferenceTimestampMs === undefined) {
-				// Log an event so we can evaluate how often we run into this scenario.
-				this.mc.logger.sendErrorEvent({
-					eventName: "GarbageCollectorInitializedWithoutTimestamp",
-					gcConfigs: JSON.stringify(this.configs),
-				});
-				return;
-			}
+			assert(
+				currentReferenceTimestampMs !== undefined,
+				"Trying to initialize GC state without current timestamp",
+			);
+
 			/**
 			 * The base snapshot data will not be present if the container is loaded from:
 			 * 1. The first summary created by the detached container.
 			 * 2. A summary that was generated with GC disabled.
 			 * 3. A summary that was generated before GC even existed.
 			 */
 			const baseSnapshotData = await this.baseSnapshotDataP;
-			if (baseSnapshotData === undefined) {
+			if (baseSnapshotData?.gcState === undefined) {
```

Comment:
You should add a 4th case in the comment above where the GC Version doesn't match, now that this early return keys off of `gcState` in particular.

Reply:
Oh, actually looks like this may be a bug/mistake.  In this case, we'll now skip the `this.summaryStateTracker.initializeBaseState` call below, but we should still be doing that if `gcState` is undefined but `baseSnapshotData` is not.

Reply:
Also do you need to do `this.gcDataFromLastRun = undefined` in this case?  Or not since this is only for initialization and not refresh.

Reply:
Oh good catch!

Reply:
> Also do you need to do this.gcDataFromLastRun = undefined in this case? Or not since this is only for initialization and not refresh.

Yeah, this only happens on initialization so its not needed anymore.